dolphin/api/v1/alert.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2020 The SODA Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from oslo_log import log
+from webob import exc
+
+from dolphin import db, cryptor
+from dolphin import exception
+from dolphin import utils
+from dolphin.api import validation
+from dolphin.api.common import wsgi
+from dolphin.api.schemas import alert as schema_alert
+from dolphin.api.views import alert as alert_view
+from dolphin.drivers import api as driver_api
+from dolphin.i18n import _
+
+LOG = log.getLogger(__name__)
+
+
+SNMPv3_keys = ('username', 'password', 'auth_enable', 'auth_protocol', 'encryption_enable',
+               'encryption_protocol', 'encryption_password', 'engine_id')
+
+
+class AlertController(wsgi.Controller):
+    def __init__(self):
+        super().__init__()
+        self.driver_api = driver_api.API()
+
+    @wsgi.response(200)
+    @validation.schema(schema_alert.put)
+    def put(self, req, id, body):
+        """Create a new alert source or update an exist one."""
+        ctx = req.environ['dolphin.context']
+        alert_source = body
+
+        try:
+            alert_source["storage_id"] = id
+            db.storage_get(ctx, id)
+            alert_source = self._input_check(alert_source)
+
+            if self._exist(ctx, id):
+                alert_source = db.alert_source_update(ctx, id, alert_source)
+            else:
+                alert_source = db.alert_source_create(ctx, alert_source)
+        except exception.StorageNotFound as e:
+            msg = (_("Alert source cannot be created or updated for a non-exist storage %s.") % id)
+            raise exc.HTTPBadRequest(explanation=msg)
+        except exception.InvalidInput as e:
+            msg = _('Failed to put alert_source: {0}'.format(e))
+            raise exc.HTTPBadRequest(explanation=msg)
```

Comment:
Any chance of getting general exception here?

Reply:
The test show, those exception will be handled in wsgi level.